### ISSUE TITLE
Simplify convergence and divergence stettings for PBS and TPS source

### ISSUE THREAD

**Issue by GitFuchs** [OpenGATE/Gate#334]
# Current status
For beam optics, the user needs to provide, sigma, divergence, emittance as well as the beam convergence (e.g. converging or diverging beam).
Due to the internal code, the sign of the divergence can not be used (squaring). Consequently, an additional parameter, beam convergence is used.

# Proposal
I propose to implement the possibility to use the sign of the divergence (negative for convergence, positive for divergence) as a default. The current convergence flags could be used as a backup to keep the macros compatible. If one of the convergence flags is set, these should take precedence.

I believe this could be implemented with minimal changes, while simplifying macros and the overal use of GATE.

**Comment by djboersma** [comment]
Thanks for bringing this up! I agree that the current implementation of convergence/divergence not ideal. One problem that I see is that for TPS, it's not possible to model a beam that is e.g. divergent at low energies and convergent at high energies. Using the sign of one of the optical parameters would possibly help to remove this limitation.

Question: I *think* that in your question you use "divergence" to mean the standard deviation of the angular distribution of the source particles, denoted by `mSigmaTheta` and `mSigmaPhi` in the code. Right? 

It's a pity that "emittance" is defined in such a weird way. Instead of emittance I'd like to use the [correlation coefficient](https://en.wikipedia.org/wiki/Pearson_correlation_coefficient) `rho` between position and angle of a particle, which would be -1 for a perfectly convergent beam, zero for a beam that is neither convergent nor divergent (which happens at the "waist" of an initially convergent beam) and +1 for a perfectly divergent beam. The non-diagonal element in the Gaussian (https://github.com/OpenGATE/Gate/blob/develop/source/physics/src/GateSourcePencilBeam.cc#L181) would then be equal to `-rho*mSigmaX*mSigmaTheta` and similar for the Y and Phi Gaussian. If the "source properties files" would provide a polynomial (or maybe we should use tables instead, like we discussed a while ago) of the correlation coefficient instead of the emittance, then your wish to encode the convergence/divergence as a sign in one of the optical parameters is perfectly implemented, I think.

Your proposal to allow the standard deviation of the angular coordinate to have a sign, positive referring to convergence and negative to divergence, is in my opinion less optimal. Because in the example above, in which a beam is divergent at low energy (`mSigmaTheta` positive) and convergent at high energy (`mSigmaTheta` negative), so then there is an energy in between at which `mSigmaTheta` will be zero (regardless of whether we implement this with polynomials like now or with tables in the near future), while actually the standard deviation of the angles is not zero at all (for this energy the particles happen to be generated at the waist of the beam).

In short:
* let's combine this change with the plan to move to tabulated (instead of polynomial) source properties
* in the tabulated source properties, the user specifies the correlation coefficient instead of the "emittance"; this allows a natural encoding of the convergence and divergence without any unphysical artifacts.

**Comment by andiresch** [comment]
Thanks for bringing this topic up again. The current situation with the flags is very bad/dangerous.

The correlation coefficient is in my opinion an excellent idea to implement a diverging/converging beam. I am not familiar with the Gate source code in this area, but is there a need to combine/mix the implementation of tabulated vs. polynomial source description and flag vs. correlation coefficient implementation? Otherwise we could do it step by step or keep both options (tabulated and polynomial).

**Comment by GrevillotL** [comment]
Hi all,

I used to write the initial code on my own as I was learning about beam optics at IBA during my PhD about 10 years ago…
I agree that different descriptions could be used, however the question is about the benefit?
If the only purpose is to remove one parameter, the gain is rather little? Or?
However I agree with David that additional features could be proposed.
I would propose the following features:

-          using a non-polynomial description would certainly be an advantage in some cases.

-          reducing the 4 parameters into 3 could be proposed as an alternative method, one could use the TWISS parameters for instance (see equations below)? ()however the definition of emittance must be very clear, as there is no consensus in accelerator physics about it, e.g emittance can be defined as: A = π.ε or A = 4.π. ε. Or other conventions… and this convention defines the interlink between the other parameters – figure 4 below)

-          In any case, please do not remove current implementation, because despite criticism, it still works well! ☺

-          In addition, I would rather consider the improvement of the TPS-Source as a whole and not only on optics. For instance calibration in number of particles per MU is currently not scaling the dose automatically. However, one could improve the code in a way that we always get the abs. dose as output, independently on the number of primaries simulated, by scaling the output dose between number of primaries simulated, prescribed primaries from the treatment plan and taking into consideration the N/MU calibration curve…


[cid:image001.jpg@01D62497.41205D30][cid:image002.jpg@01D62497.41205D30]
[cid:image003.jpg@01D62497.41205D30]


Best,
Loïc
